lib/styleToFilters.js

```diff
@@ -42,13 +45,98 @@ function styleToFilters(style) {
           layers[layerName].maxzoom = style.layers[i].maxzoom || 22;
         }
+
+        // Collect the used property
+        layers[layerName].properties = layers[layerName].properties || [];
+        ['paint', 'layout'].forEach(item => {
+          let itemObject = style.layers[i][item];
+          itemObject && getPropertyLoop(itemObject, layers[layerName].properties);
+        });
       }
     }
   }
 
+  // remove duplicate propertys and fix choose all the propertys in layers[i].properties
+  Object.keys(layers).forEach(layerId => {
+    let properties = layers[layerId].properties;
+    if (properties.indexOf(true) !== -1) {
+      layers[layerId].properties = true;
+    } else {
+      let unique = {};
+      properties.forEach(function(i) {
+        if (!unique[i]) {
+          unique[i] = true;
+        }
+      });
+      layers[layerId].properties = Object.keys(unique);
+    }
+  })
+
   return layers;
 }
 
+
+function getPropertyLoop(propertyObj, properties) {
+  Object.keys(propertyObj).forEach(key => {
+    let value = propertyObj[key];
+    // TODO we still have outher situation:
+    // - special properties: `mapbox_clip_start`, `mapbox_clip_end`
+    if (typeof value === 'string') {
+      // if the value is string try to get property name from `xx{PropertyName}xx` like.
+      // the /{[^}]+}/ig return all the value like {xxx}
+      // eg 'a{hello}badfa' => ['{hello}']
+      // eg 'a{hello}ba{world}dfa' => ['{hello}','{world}']
+      let preProperties = value.match(/{[^}]+}/ig);
+      preProperties && preProperties.forEach(item => {
+        properties.push(item.slice(1, -1));
+      });
+    } else if (typeof value === 'object' && typeof value.property === 'string') {
+      // - legacy functions with `property`
+      properties.push(value.property);
+    } else {
+      // test isExpression from sytleSpec
+      if (styleSpec.expression.isExpression(value)) {
+        // TODO: now we implement this by ourself in vtshavem, we need to talk with ‘style spec’ member to see if there have a official method to get used property, to make this can be synchronized with the expression update.
+        getPropertyFromExpression(value, properties);
+      } else {
+        // otherwise continual loop;
+        getPropertyLoop(value, properties);
+      }
+    }
+  })
+}
+
+
+function getPropertyFromExpression(exp, properties) {
+  // now we care about the expression like:
+  // ["get", string] not ["get", string, Object],
+  // ["has", string] not ["has", string, Object],
+  // ["properties"],
+  // ["feature-state", string]
+  if (exp instanceof Array) {
+    switch (exp[0]) {
+      case 'get':
+      case 'has':
+        if (typeof exp[1] === 'string' && !(exp[2] && typeof exp[2] === 'object')) {
+          properties.push(exp[1]);
+        }
+        break;
+      case 'feature-state':
+        properties.push(exp[1]);
+        break;
+      case 'properties':
+        properties.push(true);
+        break;
```

Comment:
Should the `default` case (where non of those conditions match) be handled explicitly?

Reply:
humm, in this switch, we only need to consider 4 conditions, others we will do nothing. So are we need to add the empty default like:
```
default:
        break;
```

Reply:
👍  on using `default` explicitly, if you agree. This is again, me making recommendations based on C++ guidelines and the conventions in JS may be difference, but I'm 👍 on explicitness due to http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-default. Feel free to disagree if you feel like being explicit is not idiomatic JS.

Reply:
Agree, I'll add this.